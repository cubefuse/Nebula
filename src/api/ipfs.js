import IPFS from "ipfs";
import Logger from "../util/logger";

/**
 * Used to interact with the JS-IPFS API
 */
export default class IpfsHandler {
  constructor() {
    this.node = null;
    this.createNode = this.createNode.bind(this);
    this.listDirectory = this.listDirectory.bind(this);
    this.getStats = this.getStats.bind(this);
  }

  /**
   * Starts a new IPFS node.
   * @returns {Promise<any>} A promise that resolves to a new IPFS node. Would reject on failure.
   */
  createNode() {
    Logger.debug("Starting IPFS Node", "api/ipfs");
    return new Promise((resolve, reject) => {
      this.node = new IPFS({ repo: String(Math.random() + Date.now()) });
      this.node.on("ready", () => {
        Logger.debug("IPFS node is ready", "api/ipfs");
        resolve();
      });
      this.node.on("error", () => {
        Logger.error("Error starting IPFS node", "api/ipfs");
        reject();
      });
    });
  }

  /**
   * Lists files in an IPFS directory.
   * @param path string - IPFS path string
   * @returns {Promise<Object[]>} - A promise that resolves to an array of file info objects. Would reject on failure.
   */
  listDirectory(path) {
    Logger.debug(`Getting directory info for ${path}`, "api/ipfs");
    return new Promise((resolve, reject) => {
      this.node.ls(path, (err, data) => {
        if (err) {
          Logger.debug(`Unable to get directory info for ${path}`, "api/ipfs");
          reject();
        } else {
          Logger.debug(`Received directory info for ${path}`, "api/ipfs");
          resolve(data);
        }
      });
    });
  }

  /**
   * Get statistics about the IPFS node.
   */
  getStats() {
    const node = this.node;
    async function loadStats() {
      Logger.debug(`Getting IPFS node statistics`, "api/ipfs");
      const id = node.id();
      const peers = node.swarm.peers();

      return {
        identity: await id,
        peers: await peers
      };
    }
    return loadStats();
  }
}
