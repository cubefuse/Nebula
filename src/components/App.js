import React from "react";
import { Layout } from "antd";
import { Switch } from "react-router";
import AppHeader from "./common/AppHeader";
import AppFooter from "./common/AppFooter";
import BreadcrumbsBar from "./common/BreadcrumbsBar";
import { Redirect, Route } from "react-router-dom";
import DownloadPageContainer from "../containers/DownloadPageContainer";
import FileExplorerContainer from "../containers/FileExplorerContainer";
import Error404 from "./statuses/Error404";
import Loading from "./statuses/Loading";
import Error500 from "./statuses/Error500";

const { Content } = Layout;

class App extends React.Component {
  renderRoute() {
    const routes = (
      <Switch>
        <Route exact path="/" component={FileExplorerContainer} />
        <Route path="/get/:hash" component={DownloadPageContainer} />
        <Route path="/404" component={Error404} />
        <Redirect to="/404" />
      </Switch>
    );

    switch (this.props.ipfs.connState) {
      case "offline":
        return <Loading message="Loading the app" />;
      case "connecting":
        return <Loading message="Connecting to IPFS Network" />;
      case "connected":
        return routes;
      case "failed":
        return <Error500 error="Failed to connect to IPFS." />;
      default:
        return <Error500 error="Unknown IPFS State." />;
    }
  }

  componentDidMount() {
    this.props.createIpfsInstance();
  }

  render() {
    return (
      <Layout style={{ minHeight: "100vh" }}>
        <AppHeader />
        <Content style={{ padding: "0 24px" }}>
          <BreadcrumbsBar />
          <Layout
            style={{ padding: "24px 0", background: "#fff", height: "100%" }}
          >
            {this.renderRoute()}
          </Layout>
        </Content>
        <AppFooter />
      </Layout>
    );
  }
}

export default App;
