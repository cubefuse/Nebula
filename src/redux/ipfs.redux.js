//
import { takeLatest, call, put } from "redux-saga/effects";
import IpfsHandler from "../api/ipfs";
import Logger from "../util/logger";

const IPFS = new IpfsHandler();

// State Type Definitions

// Initial State
const initialState = {
  connState: "offline"
};

// Actions
const CREATE_INSTANCE = "nebula/ipfs/CREATE_INSTANCE";
const CREATE_INSTANCE_SUCCESS = "nebula/ipfs/CREATE_INSTANCE_SUCCESS";
const CREATE_INSTANCE_FAILURE = "nebula/ipfs/CREATE_INSTANCE_FAILURE";

// Action Creators
export function createIpfsInstance() {
  return { type: CREATE_INSTANCE };
}

// Action Creator Type Definitions

// Reducer
export default function reducer(state = initialState, action) {
  switch (action.type) {
    case CREATE_INSTANCE:
      return { connState: "connecting" };
    case CREATE_INSTANCE_SUCCESS:
      return { connState: "connected" };
    case CREATE_INSTANCE_FAILURE:
      return { connState: "failed" };
    default:
      return state;
  }
}

// side effects
export function* createIpfsInstanceWatcher() {
  yield takeLatest(CREATE_INSTANCE, createIpfsInstanceWorker);
}

export function* createIpfsInstanceWorker() {
  try {
    yield call(IPFS.createNode);
    yield put({ type: CREATE_INSTANCE_SUCCESS });
  } catch (e) {
    yield put({ type: CREATE_INSTANCE_FAILURE });
  }
}
