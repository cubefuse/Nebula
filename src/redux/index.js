// @flow
import { combineReducers } from "redux";
import ipfs from "./ipfs.redux";

export default combineReducers({
  ipfs
});
