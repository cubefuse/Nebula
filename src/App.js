// @flow
import React from "react";
import { Layout } from "antd";
import { Switch } from "react-router";
import { createIpfsInstance } from "./redux/ipfs.redux";
import AppHeader from "./components/common/AppHeader";
import AppFooter from "./components/common/AppFooter";
import BreadcrumbsBar from "./components/common/BreadcrumbsBar";
import { Redirect, Route } from "react-router-dom";
import DownloadPageContainer from "./containers/DownloadPageContainer";
import FileExplorerContainer from "./containers/FileExplorerContainer";
import Error404 from "./components/statuses/Error404";
import Loading from "./components/statuses/Loading";
import Error500 from "./components/statuses/Error500";

const { Content } = Layout;

type Props = {
  dispatch: any,
  ipfs: any
};

class App extends React.Component<Props> {
  renderRoute() {
    const routes = (
      <Switch>
        <Route exact path="/" component={FileExplorerContainer} />
        <Route path="/loading" component={Loading} />
        <Route path="/get/:hash" component={DownloadPageContainer} />
        <Route path="/404" component={Error404} />
        <Redirect to="/404" />
      </Switch>
    );

    switch (this.props.ipfs.connState) {
      case "offline":
        return <Loading message="Loading the app" />;
      case "connecting":
        return <Loading message="Connecting to IPFS Network" />;
      case "connected":
        return routes;
      case "failed":
        return <Error500 error="Failed to connect to IPFS." />;
      default:
        return <Error500 error="Unknown IPFS State." />;
    }
  }

  componentDidMount() {
    this.props.dispatch(createIpfsInstance());
  }

  render() {
    return (
      <Layout style={{ minHeight: "100vh" }}>
        <AppHeader />
        <Content style={{ padding: "0 24px" }}>
          <BreadcrumbsBar />
          <Layout
            style={{ padding: "24px 0", background: "#fff", height: "100%" }}
          >
            {this.renderRoute()}
          </Layout>
        </Content>
        <AppFooter />
      </Layout>
    );
  }
}

export default App;
