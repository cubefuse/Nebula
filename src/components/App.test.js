import React from "react";
import "jest-dom/extend-expect";
import App from "./App";
import TestHelpers from "../util/test-helpers";
import { createIpfsInstance } from "../redux/ipfs.redux";

const mockIpfsState = {
  connState: "connected"
};

test("can render with mock data", () => {
  const node = TestHelpers.renderWithRouter(
    <App createIpfsInstance={jest.fn()} ipfs={mockIpfsState} />
  );
  expect(node).toMatchSnapshot();
});

test("createIpfsInstance function gets called on componentDidMount", () => {
  const mockFn = jest.fn();

  TestHelpers.renderWithRouter(<App createIpfsInstance={mockFn} ipfs={mockIpfsState} />);
  expect(mockFn).toBeCalled();
});

test("Loading the app indicator gets shown", () => {
  const mockIpfsState = { connState: "offline" };
  const node = TestHelpers.renderWithRouter(
    <App createIpfsInstance={jest.fn()} ipfs={mockIpfsState} />
  );
  expect(node).toMatchSnapshot();
});

test("Connecting to network indicator gets shown", () => {
  const mockIpfsState = { connState: "connecting" };
  const node = TestHelpers.renderWithRouter(
    <App createIpfsInstance={jest.fn()} ipfs={mockIpfsState} />
  );
  expect(node).toMatchSnapshot();
});

test("Failed to connect error gets shown", () => {
  const mockIpfsState = { connState: "failed" };
  const node = TestHelpers.renderWithRouter(
    <App createIpfsInstance={jest.fn()} ipfs={mockIpfsState} />
  );
  expect(node).toMatchSnapshot();
});
