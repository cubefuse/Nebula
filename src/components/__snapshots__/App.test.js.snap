// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connecting to network indicator gets shown 1`] = `
Object {
  "container": <div>
    <div
      class="ant-layout"
      style="min-height: 100vh;"
    >
      <div
        class="header ant-layout-header"
      >
        <div
          style="color: white;"
        >
          <span
            style="font-size: x-large; padding-right: 5px;"
          >
            <strong>
              Nebula
            </strong>
          </span>
          <span
            style="font-size: large; padding-right: 10px;"
          >
            Web UI
          </span>
          <span
            style="font-size: small;"
          >
            POC-01
          </span>
        </div>
      </div>
      <div
        class="ant-layout-content"
        style="padding: 0px 24px;"
      >
        <div
          class="ant-breadcrumb"
          style="margin: 16px 0px;"
        >
          <span>
            <span
              class="ant-breadcrumb-link"
            >
              <a
                href="#/"
              >
                Home
              </a>
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              /
            </span>
          </span>
        </div>
        <div
          class="ant-layout"
          style="padding: 24px 0px; background: rgb(255, 255, 255); height: 100%;"
        >
          <div
            class="ant-layout-content"
            style="text-align: center;"
          >
            <div
              class="ant-spin ant-spin-lg ant-spin-spinning"
            >
              <span
                class="ant-spin-dot ant-spin-dot-spin"
              >
                <i />
                <i />
                <i />
                <i />
              </span>
            </div>
            <h1
              style="font-size: xx-large; color: gray;"
            >
              Connecting to IPFS Network
            </h1>
          </div>
        </div>
      </div>
      <div
        class="ant-layout-footer"
        style="text-align: center;"
      >
        Copyright © 2018 Cubefuse. Distributed under MIT license.
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Failed to connect error gets shown 1`] = `
Object {
  "container": <div>
    <div
      class="ant-layout"
      style="min-height: 100vh;"
    >
      <div
        class="header ant-layout-header"
      >
        <div
          style="color: white;"
        >
          <span
            style="font-size: x-large; padding-right: 5px;"
          >
            <strong>
              Nebula
            </strong>
          </span>
          <span
            style="font-size: large; padding-right: 10px;"
          >
            Web UI
          </span>
          <span
            style="font-size: small;"
          >
            POC-01
          </span>
        </div>
      </div>
      <div
        class="ant-layout-content"
        style="padding: 0px 24px;"
      >
        <div
          class="ant-breadcrumb"
          style="margin: 16px 0px;"
        >
          <span>
            <span
              class="ant-breadcrumb-link"
            >
              <a
                href="#/"
              >
                Home
              </a>
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              /
            </span>
          </span>
        </div>
        <div
          class="ant-layout"
          style="padding: 24px 0px; background: rgb(255, 255, 255); height: 100%;"
        >
          <div
            class="ant-layout-content"
            style="text-align: center;"
          >
            <h1
              style="font-size: 10em; margin-bottom: 0px;"
            >
              500
            </h1>
            <h2
              style="font-size: xx-large; color: gray;"
            >
              Ouch! This application has crashed.
            </h2>
            <p>
              Debug: 
              Failed to connect to IPFS.
            </p>
          </div>
        </div>
      </div>
      <div
        class="ant-layout-footer"
        style="text-align: center;"
      >
        Copyright © 2018 Cubefuse. Distributed under MIT license.
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Loading the app indicator gets shown 1`] = `
Object {
  "container": <div>
    <div
      class="ant-layout"
      style="min-height: 100vh;"
    >
      <div
        class="header ant-layout-header"
      >
        <div
          style="color: white;"
        >
          <span
            style="font-size: x-large; padding-right: 5px;"
          >
            <strong>
              Nebula
            </strong>
          </span>
          <span
            style="font-size: large; padding-right: 10px;"
          >
            Web UI
          </span>
          <span
            style="font-size: small;"
          >
            POC-01
          </span>
        </div>
      </div>
      <div
        class="ant-layout-content"
        style="padding: 0px 24px;"
      >
        <div
          class="ant-breadcrumb"
          style="margin: 16px 0px;"
        >
          <span>
            <span
              class="ant-breadcrumb-link"
            >
              <a
                href="#/"
              >
                Home
              </a>
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              /
            </span>
          </span>
        </div>
        <div
          class="ant-layout"
          style="padding: 24px 0px; background: rgb(255, 255, 255); height: 100%;"
        >
          <div
            class="ant-layout-content"
            style="text-align: center;"
          >
            <div
              class="ant-spin ant-spin-lg ant-spin-spinning"
            >
              <span
                class="ant-spin-dot ant-spin-dot-spin"
              >
                <i />
                <i />
                <i />
                <i />
              </span>
            </div>
            <h1
              style="font-size: xx-large; color: gray;"
            >
              Loading the app
            </h1>
          </div>
        </div>
      </div>
      <div
        class="ant-layout-footer"
        style="text-align: center;"
      >
        Copyright © 2018 Cubefuse. Distributed under MIT license.
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`can render with mock data 1`] = `
Object {
  "container": <div>
    <div
      class="ant-layout"
      style="min-height: 100vh;"
    >
      <div
        class="header ant-layout-header"
      >
        <div
          style="color: white;"
        >
          <span
            style="font-size: x-large; padding-right: 5px;"
          >
            <strong>
              Nebula
            </strong>
          </span>
          <span
            style="font-size: large; padding-right: 10px;"
          >
            Web UI
          </span>
          <span
            style="font-size: small;"
          >
            POC-01
          </span>
        </div>
      </div>
      <div
        class="ant-layout-content"
        style="padding: 0px 24px;"
      >
        <div
          class="ant-breadcrumb"
          style="margin: 16px 0px;"
        >
          <span>
            <span
              class="ant-breadcrumb-link"
            >
              <a
                href="#/"
              >
                Home
              </a>
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              /
            </span>
          </span>
        </div>
        <div
          class="ant-layout ant-layout-has-sider"
          style="padding: 24px 0px; background: rgb(255, 255, 255); height: 100%;"
        >
          <div
            class="ant-layout-content"
            style="padding: 0px 24px; min-height: 280px;"
          >
            <div
              class="ant-layout-sider ant-layout-sider-dark"
              style="background: rgb(255, 255, 255); max-width: 200px; min-width: 200px; width: 200px;"
            >
              <div
                class="ant-layout-sider-children"
              >
                <ul
                  class="ant-menu ant-menu-light ant-menu-root ant-menu-inline"
                  role="menu"
                  style="height: 100%;"
                >
                  <li
                    class="ant-menu-item ant-menu-item-selected"
                    role="menuitem"
                    style="padding-left: 24px;"
                  >
                    My Files
                  </li>
                  <li
                    class="ant-menu-item"
                    role="menuitem"
                    style="padding-left: 24px;"
                  >
                    Trash
                  </li>
                  <li
                    class="ant-menu-item"
                    role="menuitem"
                    style="padding-left: 24px;"
                  >
                    Shared
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-layout-footer"
        style="text-align: center;"
      >
        Copyright © 2018 Cubefuse. Distributed under MIT license.
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
